File: javaScript/combinationSumIII.js
/*
216. Combination Sum III  
Total Accepted: 30878 Total Submissions: 86323 Difficulty: Medium

Find all possible combinations of k numbers that add up to a number n, given that only numbers from 1 to 9 can be used and each combination should be a unique set of numbers.

Ensure that numbers within the set are sorted in ascending order.

Example 1:

Input: k = 3, n = 7

Output:

[[1,2,4]]


Example 2:

Input: k = 3, n = 9

Output:

[[1,2,6], [1,3,5], [2,3,4]]

Credits:
Special thanks to @mithmatt for adding this problem and creating all test cases.
 */

/**
 * @param {number} k
 * @param {number} n
 * @return {number[][]}
 */
var combinationSum3 = function(k, n) {
	var results = [];
	var path = [];

	var dfsHelper = function(k, n, results, path, pos) {
		//base case
		if ( k === 0) {
			if (n === 0) {
				results.push(path.slice(0));				
			}
			return;
		}

		//recursive case
		for (var i = pos; i < 10; i++) {
			//pruning
			if ( i > n) {
				break;
			}

			path.push(i);
			arguments.callee(k - 1, n - i, results, path, i + 1);
			path.pop(path.lenght - 1);
		}
	}

	dfsHelper(k, n, results, path, 1);
	console.table(results);
	return results;   
};
combinationSum3(3, 9);

/*
1,2,4
*/
/*
1,2,6,1,3,5,2,3,4
*/

File: java/CombinationSumIII.java
/**
 * LeetCode 216 https://leetcode.com/problems/combination-sum-iii/
 * 
 */

class Solution {
    List<List<Integer>> res = new LinkedList<>();
    LinkedList<Integer> path = new LinkedList<>();
    
    public List<List<Integer>> combinationSum3(int k, int n) {
        backtrack(k, n, 1, 0);
        return res;
    }
    
    void backtrack(int k, int n, int startIndex, int sum) {
        // prune
        if (n < sum) {
            return;
        }
        
        //base case
        if (path.size() == k && sum == n) {
            res.add(new LinkedList(path));
            return;
        }
        
        for (int i = startIndex; i <= 9 - (k - path.size()) + 1; i++) { // another prune
            sum+= i;
            path.add(i);
            backtrack(k, n, i+1, sum);
            sum-=i;
            path.removeLast();
        }
    }
}